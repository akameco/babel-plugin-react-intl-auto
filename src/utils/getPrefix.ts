import { relative, dirname, sep } from 'path'
import { State } from '../types'
import { dotPath } from '.'

const escapeRegExp = (text: string) => {
  return text.replace(/[-[\]{}()*+?.,\\^$|#\s]/gu, '\\$&')
}

const dotPathReplace = (
  fomatted: string,
  removePrefix: string,
  separator?: string
) => {
  const exp = `^${removePrefix.replace(/\//gu, '')}\\${dotPath(
    sep,
    separator
  )}?`
  let reg: RegExp

  // certain separators can throw an error and need to be escaped
  // e.g. "_"
  try {
    reg = new RegExp(exp, 'u')
  } catch (error) {
    reg = new RegExp(escapeRegExp(exp), 'u')
  }

  return dotPath(fomatted, separator).replace(reg, '')
}

export const getPrefix = (
  {
    file: {
      opts: { filename },
    },
    opts: { removePrefix, filebase = false, separator },
  }: State,
  exportName: string | null
) => {
  if (removePrefix === true) {
    return exportName === null ? '' : exportName
  }
  const file = relative(process.cwd(), filename)
  const fomatted = filebase ? file.replace(/\..+$/u, '') : dirname(file)
  removePrefix =
    removePrefix === undefined || removePrefix === false ? '' : removePrefix

  const fixed =
    removePrefix instanceof RegExp
      ? dotPath(fomatted.replace(removePrefix, ''), separator)
      : dotPathReplace(fomatted, removePrefix, separator)

  if (exportName === null) {
    return fixed
  }

  if (fixed === '') {
    return exportName
  }

  return `${fixed}.${exportName}`
}
