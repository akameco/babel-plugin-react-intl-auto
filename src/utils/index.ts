import { sep } from 'path'
import * as t from '@babel/types'
import murmur from 'murmurhash3js'
import { NodePath } from '@babel/core'

const REG = new RegExp(`\\${sep}`, 'gu')

const isObjectProperties = (
  properties: NodePath[]
): properties is NodePath<t.ObjectProperty>[] =>
  properties.every(p => p.isObjectProperty())

export const createHash = (message: string) => `${murmur.x86.hash32(message)}`

export const dotPath = (str: string, separator: string | undefined = '.') =>
  str.replace(REG, separator)

export const objectProperty = (
  key: string,
  value: string | t.StringLiteral | t.TemplateLiteral
) => {
  const valueNode = typeof value === 'string' ? t.stringLiteral(value) : value
  return t.objectProperty(t.stringLiteral(key), valueNode)
}

export function getObjectProperties(path: NodePath) {
  if (path.isObjectExpression()) {
    const properties = path.get('properties')
    if (isObjectProperties(properties)) {
      return properties
    }
  } else if (path.isIdentifier()) {
    const binding = path.scope.getBinding(path.node.name)
    if (!binding) {
      return null
    }
    const properties = binding.path.get('init.properties')
    if (Array.isArray(properties) && isObjectProperties(properties)) {
      return properties
    }
  }
  return null
}
