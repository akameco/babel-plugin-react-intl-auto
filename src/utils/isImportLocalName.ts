import { NodePath } from '@babel/core'
import * as t from '@babel/types'
import { State } from '../types'

export const isImportLocalName = (
  name: string | null | undefined,
  allowedNames: string[],
  { file, opts: { moduleSourceName = 'react-intl' } }: State
) => {
  const isSearchedImportSpecifier = (
    specifier: NodePath<
      t.ImportDefaultSpecifier | t.ImportNamespaceSpecifier | t.ImportSpecifier
    >
  ) =>
    specifier.isImportSpecifier() &&
    allowedNames.includes(specifier.node.imported.name) &&
    (!name || specifier.node.local.name === name)

  let isImported = false

  if (file && file.path) {
    file.path.traverse({
      ImportDeclaration: {
        exit(path) {
          const specifiers = path.get('specifiers') as NodePath<
            | t.ImportDefaultSpecifier
            | t.ImportNamespaceSpecifier
            | t.ImportSpecifier
          >[]
          isImported =
            path.node.source.value.includes(moduleSourceName) &&
            specifiers.some(specifier => isSearchedImportSpecifier(specifier))

          if (isImported) {
            path.stop()
          }
        },
      },
    })
  }

  return isImported
}
