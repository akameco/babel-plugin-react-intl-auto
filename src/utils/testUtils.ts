// eslint-disable-next-line import/no-extraneous-dependencies
import pluginTester from 'babel-plugin-tester'
import { Opts } from '../types'
import plugin from '..'

export function cases(
  filename: string,
  testCases: {
    title: string
    tests: { title: string; code: string }[]
    pluginOptions?: Opts
  }[]
) {
  const defaultOpts = {
    title: '',
    plugin,
    snapshot: true,
    babelOptions: { filename, parserOpts: { plugins: ['jsx'] } },
    formatResult: (r: any) => r,
    tests: [],
  }

  for (const testCase of testCases) {
    testCase.tests = testCase.tests.map(t => ({ ...t, title: t.title }))
    pluginTester({ ...defaultOpts, ...testCase })
  }
}
