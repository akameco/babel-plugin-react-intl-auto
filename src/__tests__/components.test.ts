import path from 'path'
import { cases } from '../utils/testUtils'

const filename = path.resolve(__dirname, '..', '__fixtures__', 'messages.js')

const defaultTest = {
  title: 'default',
  code: `
import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;
`,
}

const multiUseTest = {
  title: 'multiple uses',
  code: `
import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;
<FormattedMessage defaultMessage="another" />;
`,
}

const allSupportedComponentsTest = {
  title: 'import all supported components',
  code: `
import { FormattedHTMLMessage, FormattedMessage } from 'react-intl';

<FormattedHTMLMessage defaultMessage="<span>hello</span>" />;
<FormattedMessage defaultMessage="hello" />;
`,
}

const withValueInMessageTest = {
  title: 'with a value interpolated in the message',
  code: `
import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage={\`hello world ${1 + 1}\`} />;
`,
}

const withVariableMessageTest = {
  title: 'with a variable as the defaultMessage',
  code: `
import { FormattedMessage } from 'react-intl';

const message = "variable message";

<FormattedMessage defaultMessage={message} />;
`,
}

const importAsTest = {
  title: 'with FormattedMessage imported as something else',
  code: `
import { FormattedMessage as T } from 'react-intl';

<T defaultMessage="hello" />;
`,
}

const nestedJSXTest = {
  title: 'with FormattedMessage nested in other JSX',
  code: `
import { FormattedMessage } from 'react-intl';

<div>
  <FormattedMessage defaultMessage="hello" />
</div>
`,
}

const throwWhenNotAnalyzableTest = {
  title: 'throws if defaultMessage isnâ€™t analyzable',
  code: `
import { FormattedMessage } from 'react-intl';

const getMsg = () => 'hello';

<FormattedMessage defaultMessage={getMsg()} />;
`,
  error: /\[React Intl Auto\] defaultMessage must be statically evaluate-able for extraction/u,
  snapshot: false,
}

const notTransformIfNotImportedTest = {
  title: 'does nothing if components not imported from react-intl',
  snapshot: false,
  code: `
import any from 'any-module';
<FormattedMessage defaultMessage={getMsg()} />;
`,
}

const notTransformIfSpreadAttributeTest = {
  title: 'does nothing if component props are spread',
  snapshot: false,
  code: `
import { FormattedMessage } from 'react-intl';
const props = {
  defaultMessage: 'hello'
};
<FormattedMessage {...props} />;
`,
}

const keyTest = {
  title: 'using key',
  code: `
import { FormattedMessage } from 'react-intl';

<FormattedMessage key="foobar" defaultMessage="hello" />;
`,
}

const tests = [
  defaultTest,
  multiUseTest,
  allSupportedComponentsTest,
  withValueInMessageTest,
  withVariableMessageTest,
  importAsTest,
  nestedJSXTest,
  throwWhenNotAnalyzableTest,
  notTransformIfNotImportedTest,
  notTransformIfSpreadAttributeTest,
  keyTest,
]

cases(filename, [
  { title: 'default', tests },
  {
    title: 'useKey = true',
    tests: [defaultTest, keyTest],
    pluginOptions: { useKey: true },
  },

  {
    title: 'removePrefix = "src"',
    tests: [defaultTest],
    pluginOptions: { removePrefix: 'src' },
  },

  {
    title: 'removePrefix = "src/" -- with slash',
    tests: [defaultTest],
    pluginOptions: { removePrefix: 'src/' },
  },

  {
    title: 'filebase = true',
    tests: [defaultTest],
    pluginOptions: { filebase: true },
  },

  {
    title: 'includeExportName = true',
    tests: [defaultTest],
    pluginOptions: { includeExportName: true },
  },

  {
    title: 'includeExportName = all',
    tests: [defaultTest],
    pluginOptions: { includeExportName: 'all' },
  },

  {
    title: 'removePrefix = true, includeExportName = true',
    tests: [defaultTest],
    pluginOptions: { removePrefix: true, includeExportName: true },
  },

  {
    title: 'removePrefix = false',
    tests: [defaultTest],
    pluginOptions: { removePrefix: false },
  },

  {
    title: 'removePrefix = true, includeExportName = all',
    tests: [defaultTest],
    pluginOptions: { removePrefix: true, includeExportName: 'all' },
  },

  {
    title: 'extractComments = false',
    tests: [defaultTest],
    pluginOptions: { extractComments: false },
  },

  {
    title: 'removePrefix = /__fixtures__/',
    tests: [defaultTest],
    pluginOptions: { removePrefix: /[\\/]__fixtures__/u },
  },

  {
    title: 'removePrefix = "src.__fixtures__"',
    tests: [defaultTest],
    pluginOptions: { removePrefix: 'src.__fixtures__' },
  },

  {
    title: 'removeDefaultMessage = true',
    tests: [defaultTest],
    pluginOptions: { removeDefaultMessage: true },
  },
])
