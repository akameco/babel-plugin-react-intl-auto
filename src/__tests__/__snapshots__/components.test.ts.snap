// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`default import all supported components: import all supported components 1`] = `

import { FormattedHTMLMessage, FormattedMessage } from 'react-intl';

<FormattedHTMLMessage defaultMessage="<span>hello</span>" />;
<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedHTMLMessage, FormattedMessage } from 'react-intl';
<FormattedHTMLMessage id="src.__fixtures__.205247277" defaultMessage="<span>hello</span>" />;
<FormattedMessage id="src.__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`default multiple uses: multiple uses 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;
<FormattedMessage defaultMessage="another" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.__fixtures__.613153351" defaultMessage="hello" />;
<FormattedMessage id="src.__fixtures__.1418522856" defaultMessage="another" />;

`;

exports[`default using key: using key 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage key="foobar" defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage key="foobar" id="src.__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`default with FormattedMessage imported as something else: with FormattedMessage imported as something else 1`] = `

import { FormattedMessage as T } from 'react-intl';

<T defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage as T } from 'react-intl';
<T id="src.__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`default with FormattedMessage nested in other JSX: with FormattedMessage nested in other JSX 1`] = `

import { FormattedMessage } from 'react-intl';

<div>
  <FormattedMessage defaultMessage="hello" />
</div>

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<div>
  <FormattedMessage id="src.__fixtures__.613153351" defaultMessage="hello" />
</div>;

`;

exports[`default with a value interpolated in the message: with a value interpolated in the message 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage={\`hello world 2\`} />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.__fixtures__.3540954520" defaultMessage={\`hello world 2\`} />;

`;

exports[`default with a variable as the defaultMessage: with a variable as the defaultMessage 1`] = `

import { FormattedMessage } from 'react-intl';

const message = "variable message";

<FormattedMessage defaultMessage={message} />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
const message = "variable message";
<FormattedMessage id="src.__fixtures__.3082794952" defaultMessage={message} />;

`;

exports[`extractComments = false default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`filebase = true default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.__fixtures__.messages.613153351" defaultMessage="hello" />;

`;

exports[`includeExportName = all default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`includeExportName = true default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`removeDefaultMessage = true default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.__fixtures__.613153351" />;

`;

exports[`removePrefix = "src" default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`removePrefix = "src.__fixtures__" default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="613153351" defaultMessage="hello" />;

`;

exports[`removePrefix = "src/" -- with slash default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`removePrefix = /__fixtures__/ default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.613153351" defaultMessage="hello" />;

`;

exports[`removePrefix = false default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`removePrefix = true, includeExportName = all default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="613153351" defaultMessage="hello" />;

`;

exports[`removePrefix = true, includeExportName = true default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="613153351" defaultMessage="hello" />;

`;

exports[`useKey = true default: default 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage id="src.__fixtures__.613153351" defaultMessage="hello" />;

`;

exports[`useKey = true using key: using key 1`] = `

import { FormattedMessage } from 'react-intl';

<FormattedMessage key="foobar" defaultMessage="hello" />;

      ↓ ↓ ↓ ↓ ↓ ↓

import { FormattedMessage } from 'react-intl';
<FormattedMessage key="foobar" id="src.__fixtures__.foobar" defaultMessage="hello" />;

`;
